feat: added better support for keyboard menu navigation

preda-bogdan opened the pull request:
### Summary
<!-- Please describe the changes you made. -->
Added better support for keyboard navigation.
I moved the sidebar animation inline and is enqueued only if sidebar navigation is used.
Fixed JS trap for sidebar to allow removing trap on menu close. Previously was only working on ESC.

For mega menu use this PR for Neve Pro:
https://github.com/Codeinwp/neve-pro-addon/pull/2268

Previous behavior:
Before it would open the sub-menu automatically on parent TAB.
Then it would TAB through all children before moving to the next item.

Current behavior:
Now the caret is a focus-able element and will accept ENTER as a key to toggle the sub-menu
TAB only cycles through visible/opened items.
If the sub-menu is not opened from the caret, the TAB will move to next item.


### Will affect visual aspect of the product
<!-- It includes visual changes? -->
NO

### Screenshots <!-- if applicable -->
https://d.pr/v/NUoLv6


### Test instructions
<!-- Describe how this pull request can be tested. -->
1. Check with a menu that has sub-menus 
2. Use the keyboard to navigate
3. Ensure that all previous functionality behaves the same.

<!-- Issues that this pull request closes. -->
Closes #2935.
Closes #3731.
<!-- Should look like this: `Closes #1, #2, #3.` . -->

preda-bogdan commented:
**Notes:**
I tried to keep under the size limit for the CSS and JS changes.
However I do inline some JS that is required for the caret to toggle the sub-menu and for the sub-menu to disable the active state of the caret when losing focus. I could not find a better way of doing this but I am opened to suggestions if you have any.

abaicus commented on inc/views/nav_walker.php:
```diff
@@ -172,14 +179,17 @@ private function get_caret_pictogram( $settings ) {
 		$mobile_button_caret_css .= '.header-menu-sidebar .nav-ul li .wrap a .dd-title { width: var(--wrapdropdownwidth); }';
 		$mobile_button_caret_css .= '.header-menu-sidebar .nav-ul li .wrap button { border: 0; z-index: 1; background: 0; }';
-		return Dynamic_Css::minify_css( $mobile_button_caret_css );
+		$accessibility_caret_css  = '.nav-ul:not(.menu-mobile) li { display: flex; align-items: center; }';
```
@preda-bogdan I am pretty sure this line here will break both the free & pro **mega-menu** features.

We should make sure that it still works fine even after the changes.

preda-bogdan replied:
@abaicus thank you for mentioning it, I managed to find a solution that works on both. Let me know what you think.

abaicus reviewed (changes requested):
@preda-bogdan, while the code looks good, I am pretty sure some changes here will break the mega-menu features.

abaicus commented on inc/views/nav_walker.php:
```diff
@@ -55,7 +62,7 @@ public function inline_style_for_sidebar() {
 			return;
 		}
-		echo '<style>' . wp_kses_post( $this->get_mobile_submenu_style() ) . '</style>';
+		echo '<style>' . $this->get_mobile_submenu_and_accessibility_style() . '</style>'; //phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
Would wp_kses_post alter the inline styles added here?

preda-bogdan replied:
Yes, it would break `>` direct descendent referencing, I could not keep that. But since we have control over the style content I think we can skip escaping that since we don't use any filters and we know the content.

abaicus replied:
Makes sense, @preda-bogdan! 👍🏻 Just wanted to check 🚀

abaicus reviewed (approved):
Looking good! 👍🏻 I just added a question regarding sanitization.

Other than that, here are a few additional things that I think we should probably test in QA:

- Expanded first-level menu items setting in customizer for the sidebar;
- Both free and pro mega-menu features;
- The hover effects for the primary menu;

irinelenache commented:
@preda-bogdan Tested the PR and here are my findings:

- [ ] The icons in the mega menu have different positions after using this PR https://vertis.d.pr/i/nNO1GB
- [ ] The elements in the mega menu aren't skipped if i don't open the dropdown, i have to press tab until i iterate through all of them https://vertis.d.pr/v/gmWbwx
- [ ] If the submenu icon position is set to **Left**, opening the dropdowns won't work anymore https://vertis.d.pr/v/2UfgHE
- [ ] Submenu icon doesn't work anymore if it's set to be an image https://vertis.d.pr/i/8WpWJ5
- [ ] The hover skin mode is applied only to the menu item text, instead of the menu item text + icon https://vertis.d.pr/i/o4vNLb
- [ ] The icon position changed is different now
before the PR: https://vertis.d.pr/i/vxNQDB
with the PR: https://vertis.d.pr/i/o4vNLb
- [ ] This is not related to this PR, i just discovered it now: In the menu sidebar, the navigation should be circular and the user should close it using the X button. If the mega menu is present inside the sidebar, the circular navigation will work only if the mega menu is expanded. If it's not expanded, the sidebar will remain open and the focus will switch to the next items on the page. Here is a video with the behaviour https://vertis.d.pr/v/PMzAfO

You can check the issues here:
```
URL : https://unk-mayhem-yoga.vertisite.cloud
Username : mega-menu
Password : wC1ty0KfqPNQ6enDIYsh
```

preda-bogdan commented:
@irinelenache when you have some time please take a second look at this I submitted changes to solve all the issues you mentioned. Thank you!

irinelenache commented:
@preda-bogdan Thank you 🚀  Checked again and some issues were not entirely fixed for me, here is what i found:

>  The elements in the mega menu aren't skipped if i don't open the dropdown, i have to press tab until i iterate through all of them https://vertis.d.pr/v/gmWbwx

^ This one is still happening for me

>  If the submenu icon position is set to Left, opening the dropdowns won't work anymore https://vertis.d.pr/v/2UfgHE

^ Regarding this issue, this is how it works now [using the mouse](https://vertis.d.pr/i/GJ4DFx) and this is how it works [using the keyboard navigation](https://vertis.d.pr/v/rnOKRl) 

> This is not related to this PR, i just discovered it now: In the menu sidebar, the navigation should be circular and the user should close it using the X button. If the mega menu is present inside the sidebar, the circular navigation will work only if the mega menu is expanded. If it's not expanded, the sidebar will remain open and the focus will switch to the next items on the page. Here is a video with the behaviour https://vertis.d.pr/v/PMzAfO

^ Regarding this issue, now the navigation is circular, but after closing the menu sidebar, i can't iterate further than the next element after the menu icon https://vertis.d.pr/v/iy4ANo

- [ ] The mega menu icons are not aligned https://vertis.d.pr/i/tWd8SJ

- [ ] In the sidebar menu, the hover skin mode is applied to more subitems than it should https://vertis.d.pr/v/uNIyrF

The issues can be checked on the same instance 🙏

preda-bogdan commented:
@irinelenache Thank you for checking this. 💪🏼  There were still some issues present when changing the menu styles. Thank you for catching those. I updated the instance and fixed the issues. There is also a PR from PRO that is required now for the mega menu: https://github.com/Codeinwp/neve-pro-addon/pull/2268

Please have another look when you have some time and let me know if you find anything else or if it is working as expected now. Thank you! 🙇🏼‍♂️
